File: src/main/java/proto/games/matchmaker/domain/matchmaker/rank/RankTeam.java
package proto.games.matchmaker.domain.matchmaker.rank;

import lombok.Getter;
import proto.games.matchmaker.domain.player.Rank;

import java.util.ArrayList;
import java.util.stream.Collectors;

@Getter
public class RankTeam extends ArrayList<Rank> {
    private final int limit;

    public RankTeam(Rank baseRank) {
        limit = baseRank.getWeight() * 5;
    }

    public int getCurrentWeight() {
        return this.stream()
                .parallel()
                .collect(Collectors.summingInt(Rank::getWeight));
    }

    public int getRemainingSpace() {
        return limit - getCurrentWeight();
    }
}

File: src/main/kotlin/proto/games/matchmaker/domain/matchmaker/rank/RankTeam.kt
package proto.games.matchmaker.domain.matchmaker.rank

import proto.games.matchmaker.domain.player.Rank

class RankTeam(baseRank: Rank) : ArrayList<Rank>() {
    val limit: Int = baseRank.weight * 5
    val currentWeight: Int
        get() = sumBy { it.weight }
    val remainingSpace: Int
        get() = limit - currentWeight

}